backend/src/packages/orders/order.service.ts

```diff
@@ -0,0 +1,249 @@
+import { type IService } from '~/libs/interfaces/interfaces.js';
+
+import { type BusinessService } from '../business/business.service.js';
+import { type DriverService } from '../drivers/driver.service.js';
+import { type UserEntityObjectWithGroupT } from '../users/users.js';
+import { OrderStatus, UserGroupKey } from './libs/enums/enums.js';
+import {
+  type OrderCreateRequestDto,
+  type OrderCreateResponseDto,
+  type OrderEntity as OrderEntityT,
+  type OrderFindByIdResponseDto,
+  type OrderUpdateRequestDto,
+  type OrderUpdateResponseDto,
+} from './libs/types/types.js';
+import { OrderEntity } from './order.entity.js';
+import { type OrderRepository } from './order.repository.js';
+
+class OrderService implements Omit<IService, 'find'> {
+  private orderRepository: OrderRepository;
+
+  private businessService: BusinessService;
+
+  private driverService: DriverService;
+
+  public constructor({
+    businessService,
+    orderRepository,
+    driverService,
+  }: {
+    orderRepository: OrderRepository;
+    businessService: BusinessService;
+    driverService: DriverService;
+  }) {
+    this.orderRepository = orderRepository;
+
+    this.businessService = businessService;
+
+    this.driverService = driverService;
+  }
+
+  public async create(
+    payload: OrderCreateRequestDto & {
+      userId: number | null;
+    },
+  ): Promise<OrderCreateResponseDto> {
+    const {
+      scheduledTime,
+      carsQty,
+      startPoint,
+      endPoint,
+      customerName,
+      customerPhone,
+      driverId,
+      userId,
+    } = payload;
+    const driver = await this.driverService.findById(driverId);
+
+    if (!driver) {
+      throw new NotFoundError({ message: HttpMessage.DRIVER_DOES_NOT_EXIST });
+    }
+    const order = await this.orderRepository.create(
+      OrderEntity.initializeNew({
+        price: 100, //Mock, get price from truck and calculated distance
+        scheduledTime,
+        carsQty,
+        startPoint,
+        endPoint,
+        status: OrderStatus.PENDING,
+        userId,
+        businessId: driver.businessId,
+        driverId,
+        customerName,
+        customerPhone,
+      }),
+    );
+
+    return order.toObject();
+  }
+
+  public async findById(id: OrderEntityT['id']): Promise<OrderEntityT | null> {
+    const foundOrder = await this.orderRepository.findById(id);
+
+    if (!foundOrder) {
+      throw new NotFoundError({
+        message: HttpMessage.ORDER_DOES_NOT_EXIST,
+      });
+    }
+
+    return foundOrder.toObject();
+  }
+
+  public async update(parameters: {
+    id: OrderEntityT['id'];
+    payload: OrderUpdateRequestDto;
+  }): Promise<OrderUpdateResponseDto> {
+    const updatedOrder = await this.orderRepository.update(parameters);
+
+    if (!updatedOrder) {
+      throw new NotFoundError({
+        message: HttpMessage.ORDER_DOES_NOT_EXIST,
+      });
+    }
+
+    return updatedOrder.toObject();
+  }
+
+  public async findOne({
```

Comment:
```suggestion
  const result = await this.orderRepository.findByIdWithDriver(id);
  if (!result) {
    throw new NotFoundError({message: HttpMessage.ORDER_DOES_NOT_EXIST,});
  }

  const { order, driver } = result;
  //do we need create foundOrder var, or can we reuse result found from the repo
  const foundOrder = { ...order.toObject(), driver };

  if (user?.group.key === UserGroupKey.BUSINESS) {
    const business = await this.businessService.findByOwnerId(user.id);
    this.verifyOrderBelongsToBusiness(foundOrder, business);
}
  if (user) {
    this.verifyOrderBelongsToUser(foundOrder, user);
  }

  // how do we verify that user that is not logged can view this order?
  return foundOrder;
```

Reply:
//do we need create foundOrder var, or can we reuse result found from the repo
No we can't reuse result from repo because entity of order, not include driver, and we can't add driver to order entity because when we create order we should not add driver to it, So we can't use `order.toObject()` to entity returned from repo
 // how do we verify that user that is not logged can view this order?
we can't verify it for now, because we need some kind of token to identify unregistered user